### PR TITLE
add an 'equals' method to ItemStack and compatibility w/ lua '=='

### PR DESCRIPTION
**Pull request by fluxionary**
adds an `equals` method to `ItemStack`. also supports using the `==` operator.

## How to test

this is what i used to test this:

```lua
local i1 = ItemStack("default:stone")
local i2 = ItemStack("default:stone")
local i3 = ItemStack("default:stone")

local m1 = i1:get_meta()
local m2 = i2:get_meta()

local keys = {"a", "b", "c", "d", "e", "f", "g", "h", "i", "j", "k", "l", "m", "n", "o", "p"}
local values = {}

for i, key in pairs(keys) do
 m1:set_int(key, i)
 values[key] = i
end

for key, i in pairs(values) do
 m2:set_int(key, i)
end

assert(i1:equals(i2))
assert(i1 == i2)
assert(not i1:equals(i3))
assert(i1 ~= i3)
```

### CODE REVIEW

**Comment by appgurueu** [comment]
> this is what i used to test this

Why not add this to the unit tests?

**Comment by fluxionary** [comment]
> > this is what i used to test this
> 
> Why not add this to the [unit tests](https://github.com/minetest/minetest/tree/master/games/devtest/mods/unittests)?

oh *that's* where they are. i looked and was surprised i didn't find them, will do.

**Comment by Desour** [comment]
* How does `item1 == item2` compare to `item1.to_string() == item2.to_string()`?
  (The answer to this would also be useful in doc.)
* Metamethod `==` is missing in documentation.

**Comment by fluxionary** [comment]
> * How does `item1 == item2` compare to `item1.to_string() == item2.to_string()`?
>       (The answer to this would also be useful in doc.)

`item1 == item2` may return true when `item1.to_string() == item2.to_string()` returns false, due to order in which metadata is serialized. 

>     * Metamethod `==` is missing in documentation.

it's also missing for MetaDataRef, i could add that if it's appropriate.

**Comment by Zughy** [comment]
@fluxionary please rebase

File: doc/lua_api.txt [hunk 1]
```diff
@@ -6763,6 +6763,18 @@ an itemstring, a table or `nil`.
 * `peek_item(n)`: returns taken `ItemStack`
     * Copy (don't remove) up to `n` items from this stack
     * `n`: number, default: `1`
+* `equals(other)`:
+    * returns `true` if this stack is identical to `other`.
+    * Note: `stack1:to_string() == stack2:to_string()` is not reliable,
+      as stack metadata can be serialized in arbitrary order.
+
+Operators
```
Comment (TurkeyMcMac) [inline doc/lua_api.txt:6771]: Shouldn't this be `### Operators` so that it is placed in the `ItemStack` section?

**Review by TurkeyMcMac — COMMENTED** [review]
Looks good except for one thing.

File: doc/lua_api.txt (hunk 1 above)
Comment (fluxionary) [reply doc/lua_api.txt:6771]: i dunno; i'm borrowing the syntax for an `operators` section from https://github.com/minetest/minetest/blob/500ebc859d90ad51e366530d8ca01ab0cfe5425f/doc/lua_api.txt#L3511-L3512  

easy to fix, if someone confirms way it should be.

Comment (TurkeyMcMac) [reply doc/lua_api.txt:6771]: There "Operators" (header level 2) is a subsection of "Spatial Vectors" (header level 1). Here "Operators" should be header level 3, as it is a subsection of "`ItemStack`" (header level 2).

**Review by TurkeyMcMac — APPROVED** [review]
LGTM

**Comment by rubenwardy** [comment]
The following code causes a crash:

```lua
ItemStack("basenodes:stone") == { foo = "bar"}
```

it should instead return false. I've added a unit test for this

**Comment by TurkeyMcMac** [comment]
I think it's only `ItemStack("basenodes:stone"):equals({ foo = "bar"})` that fails. For `==`, Lua always returns false for values of different types or different `__eq` metamethods.

**Comment by TurkeyMcMac** [comment]
`MetaDataRef::l_equals` also expects both arguments to be metadata objects.

**Comment by rubenwardy** [comment]
You're right. In which case, `==` and `~=` have the correct behaviour. The `:equals()` method should either be changed to match `==`, or it requiring an ItemStack should be documented

**Comment by TurkeyMcMac** [comment]
IMO for now the limitation should just be documented, since right now it's consistent with `MetaDataRef:equals` and making the method accept any value would increase the code complexity significantly (`luaL_checkudata` would have to be partially reimplemented.)

File: src/script/lua_api/l_item.cpp [hunk 2]
```diff
@@ -437,8 +437,32 @@ int LuaItemStack::l_peek_item(lua_State *L)
 int LuaItemStack::l_equals(lua_State *L)
 {
 	LuaItemStack *o1 = checkobject(L, 1);
-	ItemStack &item1 = o1->m_stack;
+
+	void *p = lua_touserdata(L, 2);
```
Comment (sfan5) [inline src/script/lua_api/l_item.cpp:441]: I think `lua_type(L, 2) == LUA_TUSERDATA` should work as well
also there should be a (short) comment here explaining why it does this.

File: src/script/lua_api/l_item.cpp [hunk 3]
```diff
@@ -437,8 +437,32 @@ int LuaItemStack::l_peek_item(lua_State *L)
...
+
+	lua_getfield(L, LUA_REGISTRYINDEX, className);
+	if (!lua_rawequal(L, -1, -2)) {
+		lua_pushboolean(L, false);
+		return 1;
+	}
+
 	LuaItemStack *o2 = checkobject(L, 2);
+
+	if (o1 == NULL || o2 == NULL) {
```
Comment (sfan5) [inline src/script/lua_api/l_item.cpp:460]: checkobject never returns NULL

Comment (fluxionary) [reply src/script/lua_api/l_item.cpp:461]: right, that's a leftover from before i figured out how to do this correctly, will remove.

File: src/script/lua_api/l_item.cpp (hunk 2 above)
Comment (fluxionary) [reply src/script/lua_api/l_item.cpp]: seems reasonable, i based the code on `luaL_checkudata`. there's also apparently a `lua_isuserdata`, would that be better than your proposal?

Comment (sfan5) [reply src/script/lua_api/l_item.cpp]: works too I guess

File: src/script/lua_api/l_item.cpp [hunk 4]
```diff
@@ -433,6 +433,40 @@ int LuaItemStack::l_peek_item(lua_State *L)
...
+		lua_pushboolean(L, false);
+		return 1;
+	}
+
+ 	// checks for non-ItemStack argument
+	lua_getfield(L, LUA_REGISTRYINDEX, className);
+	if (!lua_rawequal(L, -1, -2)) {
+		lua_pushboolean(L, false);
+		return 1;
+	}
```
Comment (sfan5) [inline src/script/lua_api/l_item.cpp:459]:
```suggestion
 	// checks that the argument is an ItemStack
	if (!lua_getmetatable(L, 2)) {
		lua_pushboolean(L, false);
		return 1;
	}
	lua_getfield(L, LUA_REGISTRYINDEX, className);
	if (!lua_rawequal(L, -1, -2)) {
		lua_pushboolean(L, false);
		return 1;
	}
```

**Review by sfan5 — APPROVED** [review]

**Pull request merged**